In [ ]:
# Dounloading the Dataset
!wget --header="Host: storage.googleapis.com" --header="User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.81 Safari/537.36" --header="Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9" --header="Accept-Language: en-GB,en-US;q=0.9,en;q=0.8" --header="Referer: https://www.kaggle.com/" "https://storage.googleapis.com/kaggle-data-sets/1582264/2603511/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20220209%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20220209T112913Z&X-Goog-Expires=259199&X-Goog-SignedHeaders=host&X-Goog-Signature=44b3e18fc9f08d18f8665050cc2f95350d90d0d3e699e295a02ec312945f8ccdedd3877010849de40b43941cb1f327cc179811ec0cf4d16775a0a416e36565e52a408daf6e7bf9eb4671c89991709d38b289c61d4342c050ce9862f81619e4fe8cd2a9cbbc11835060a58a37cdd489fe42fd0219f329a2fb3e0717ae5758fcec2aa8d39bb320857c9437b63e1904c25c130c107c3376d59662fe29d2fa3ca7f7d3cdb00271223b040902fea79cdbd7d3667008275dacb369a38a46b7c947eedaae40c973c9a14cd34fcac2bc8d9b9ab11698184e9a33d5b1236f57cc901179b5fdc5aa1c779e064271fed77f5e00af7cd360054550fc1663f0ba3e68dc05c7c5" -c -O 'archive.zip'
!unzip archive.zip

In [1]:
# Importing the Required Libraries
import tensorflow as tf
import keras
import numpy as np
from sys import getsizeof
from tensorflow.keras.layers import *
import matplotlib.pyplot as plt
from tensorflow.keras.preprocessing.image import ImageDataGenerator
import os
from sklearn.metrics import accuracy_score

In [ ]:
def get_size(file_path):
  """"Prints the size of the file in MB"""
  size = os.path.getsize(file_path)

  print('File size: ' + str(round(size/(1024 * 1024),3)) + ' Megabytes')    

In [ ]:
# Initializing and loading the data using ImageDataGenerator
test_datagen = ImageDataGenerator(rescale=1./255)

validation_generator = test_datagen.flow_from_directory(
    'DevanagariHandwrittenCharacterDataset/Test',
    target_size=(32, 32),
    batch_size=64,
    seed = 7,
    class_mode='categorical',
    color_mode='grayscale')

Found 13800 images belonging to 46 classes.


In [ ]:
# Loading the trained TF Model
Model = tf.keras.models.load_model('/content/drive/MyDrive/CS2_Model1.h5')

In [ ]:
# Evaluating the Model performance
Model.evaluate(validation_generator)

216/216 [==============================] - 29s 130ms/step - loss: 0.1285 - accuracy: 0.9606


[0.12848027050495148, 0.9605796933174133]

In [ ]:
# Printing the Accuracy of the Model
print('The Accuracy of the Model is 96.0%')

The Accuracy of the Model is 96.0%


In [ ]:
# Printing the size of the File
get_size('/content/drive/MyDrive/CS2_Model1.h5')

File size: 150.4 Megabytes


# TF Lite Model

In [ ]:
# Converting the Tensorflow Model to Tensorflow Lite Model
tf_lite_converter = tf.lite.TFLiteConverter.from_keras_model(Model)
tf_lite_converter.optimizations = [tf.lite.Optimize.OPTIMIZE_FOR_SIZE]

TF_Lite_Model = tf_lite_converter.convert()

INFO:tensorflow:Assets written to: /tmp/tmp3yvouc66/assets


INFO:tensorflow:Assets written to: /tmp/tmp3yvouc66/assets


In [ ]:
# Saving the Tensorflow Lite model in "TF_Lite_Model.tflite" file
open('TF_Lite_Model.tflite','wb').write(TF_Lite_Model)

13149456

In [ ]:
# Printing the size of the TF Lite model
get_size('TF_Lite_Model.tflite')

File size: 12.54 Megabytes


In [ ]:
# Checking the Input and Output shapes 
interpreter = tf.lite.Interpreter(model_path='TF_Lite_Model.tflite')
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
print('The Input Shape is ', input_details[0]['shape'])
print('The Input Data Type is ', input_details[0]['dtype'])
print('The Output Shape is ', output_details[0]['shape'])
print('The Output Data Type is ', output_details[0]['dtype'])

The Input Shape is  [ 1 32 32  1]
The Input Data Type is  <class 'numpy.float32'>
The Output Shape is  [ 1 46]
The Output Data Type is  <class 'numpy.float32'>


In [ ]:
# Resizing the input and output shapes 
interpreter.resize_tensor_input(input_details[0]['index'], (64, 32, 32, 1))
interpreter.resize_tensor_input(output_details[0]['index'], (64, 46))

In [ ]:
# Checking the Input and Output shapes after resizing
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
print('The Input Shape is ', input_details[0]['shape'])
print('The Input Data Type is ', input_details[0]['dtype'])
print('The Output Shape is ', output_details[0]['shape'])
print('The Output Data Type is ', output_details[0]['dtype'])

The Input Shape is  [64 32 32  1]
The Input Data Type is  <class 'numpy.float32'>
The Output Shape is  [64 46]
The Output Data Type is  <class 'numpy.float32'>


In [ ]:
# Evaluating the TF Lite model Performance
actual_class = []
predicted_class = []

for data, label in validation_generator:
  if (len(label) < validation_generator.batch_size) and (len(predicted_class) >= 13700):
    break

  actual_class.extend(np.argmax(label,axis=-1))
  interpreter.allocate_tensors()
  interpreter.set_tensor(input_details[0]['index'], data)
  interpreter.invoke()
  tflite_model_predictions = interpreter.get_tensor(output_details[0]['index'])
  prediction_classes = np.argmax(tflite_model_predictions, axis=1)
  predicted_class.extend(prediction_classes.tolist())

    # We have to break out from the generator when we've processed 
    # the entire once (otherwise we would end up with duplicates). 
  if (len(label) < validation_generator.batch_size) and :
      break

In [ ]:
# Printing the Accuracy of the TF Lite Model
accuracy = accuracy_score(predicted_class, actual_class)
print('The Accuracy of the TF Lite Model is {:.1f}%'.format(accuracy*100))

The Accuracy of the TF Lite Model is 96.2%


By converting our model from Tensorflow to Tensorflow Lite we could decrease the size of the file from 150 MB to 12 MB with no drop in the performance of the model